### PR TITLE
fix: Fix logs part 2

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the USERS.md.
* [ ] I've signed the CLA and required builds are green. 

See #2589

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -210,20 +214,23 @@ endif
...
+dist/update-mocks: $(HOME)/go/bin/mockery $(MOCK_FILES)
+	./hack/update-mocks.sh $(MOCK_FILES)
+	touch dist/update-mocks
+
 .PHONY: codegen
-codegen: $(HOME)/go/bin/mockery
-	# Generate code
+codegen: dist/update-mocks
+	# Back-up go.*, but only if we have not already done this (because that would suggest we failed mid-codegen and the currenty go.* files are borked).
+	[[ -e dist/go.mod ]] || cp go.mod go.sum dist/
```
Comment (alexec) [inline Makefile:224]: this change is intended to stop the codegen job being red due to golang/tools changes

File: Makefile [hunk 2]
```diff
@@ -81,11 +81,15 @@ MANIFESTS        := $(shell find manifests          -mindepth 2 -type f)
...
 # The sort puts _.primary first in the list. 'env LC_COLLATE=C' makes sure underscore comes first in both Mac and Linux.
-SWAGGER_FILES    := $(shell find pkg -name '*.swagger.json' | env LC_COLLATE=C sort)
+SWAGGER_FILES    := $(shell find pkg/apiclient -name '*.swagger.json' | env LC_COLLATE=C sort)
+MOCK_FILES       := $(shell find . -maxdepth 4 -not -path '*/vendor/*' -path '*/mocks/*' -type f)
 
 .PHONY: build
 build: status clis executor-image controller-image manifests/install.yaml manifests/namespace-install.yaml manifests/quick-start-postgres.yaml manifests/quick-start-mysql.yaml
 
+# https://stackoverflow.com/questions/4122831/disable-make-builtin-rules-and-variables-from-inside-the-make-file
+.SUFFIXES:
```
Comment (alexec) [inline Makefile:91]: this makes it a tiny bit faster, but also makes `make -d` output much shorter

File: Makefile [hunk 3]
```diff
@@ -210,20 +214,23 @@ endif
 $(HOME)/go/bin/mockery:
 	go get github.com/vektra/mockery/.../
 
+dist/update-mocks: $(HOME)/go/bin/mockery $(MOCK_FILES)
```
Comment (alexec) [inline Makefile:217]: de-couple the mocks from other codegen, ultimately I'd like codegen to be smart enough to only run if needed

File: util/logs/workflow-logger.go [hunk 4]
```diff
@@ -106,7 +103,7 @@ func NewWorkflowLogger(wfClient versioned.Interface, kubeClient kubernetes.Inter
 				scanner := bufio.NewScanner(stream)
 				for scanner.Scan() {
 					select {
-					case <-stopChan:
+					case <-ctx.Done():
```
Comment (alexec) [inline util/logs/workflow-logger.go:106]: we were stopping the output as soon as the workflow complete, not only was this unnecessary it was stopping the logs too soon

File: Makefile (hunk 3 above)
Comment (simster7) [reply Makefile:217]: Nice!

**Pull request merged**